python/pyspark/sql/streaming/stateful_processor.py

```diff
@@ -0,0 +1,160 @@
+from pyspark.sql.types import StructType, _parse_datatype_string
+
+if TYPE_CHECKING:
+    from pyspark.sql.pandas._typing import DataFrameLike as PandasDataFrameLike
+
+__all__ = ["StatefulProcessor", "StatefulProcessorHandle"]
+
+
+class ValueState:
+    """
+    Class used for arbitrary stateful operations with the v2 API to capture single value state.
+
+    .. versionadded:: 4.0.0
+    """
+
+    def __init__(
+        self, value_state_client: ValueStateClient, state_name: str, schema: Union[StructType, str]
+    ) -> None:
+        self._value_state_client = value_state_client
+        self._state_name = state_name
+        self.schema = schema
+
+    def exists(self) -> bool:
+        """
+        Whether state exists or not.
+        """
+        return self._value_state_client.exists(self._state_name)
+
+    def get(self) -> Any:
+        import pandas as pd
+
+        """
+        Get the state value if it exists.
+        """
+        value = self._value_state_client.get(self._state_name)
+        schema = self.schema
+        if isinstance(schema, str):
+            schema = cast(StructType, _parse_datatype_string(schema))
+        columns = [field.name for field in schema.fields]
+        # Create the DataFrame using the values and schema
+        df = pd.DataFrame([value], columns=columns)
+        return df
+
+    def update(self, new_value: Any) -> None:
+        """
+        Update the value of the state.
+        """
+        self._value_state_client.update(self._state_name, self.schema, new_value)
+
+    def clear(self) -> None:
+        """
+        Remove this state.
+        """
+        self._value_state_client.clear(self._state_name)
+
+
+class StatefulProcessorHandle:
+    """
+    Represents the operation handle provided to the stateful processor used in the arbitrary state
+    API v2.
+
+    .. versionadded:: 4.0.0
+    """
+
+    def __init__(self, stateful_processor_api_client: StatefulProcessorApiClient) -> None:
+        self.stateful_processor_api_client = stateful_processor_api_client
+
+    def getValueState(self, state_name: str, schema: Union[StructType, str]) -> ValueState:
+        """
+        Function to create new or return existing single value state variable of given type.
+        The user must ensure to call this function only within the `init()` method of the
+        StatefulProcessor.
+
+        Parameters
+        ----------
+        state_name : str
+            name of the state variable
+        schema : :class:`pyspark.sql.types.DataType` or str
+            The schema of the state variable. The value can be either a
+            :class:`pyspark.sql.types.DataType` object or a DDL-formatted type string.
+        """
+        self.stateful_processor_api_client.get_value_state(state_name, schema)
+        return ValueState(ValueStateClient(self.stateful_processor_api_client), state_name, schema)
+
+
+class StatefulProcessor(ABC):
```

Comment:
Has to be listed in `python/docs/source/reference/pyspark.sql` so it can be documented.